src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -763,6 +763,78 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                     preserveSig);
             }
         }
+
+        private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is object);
+
+            if (MethodKind != MethodKind.Ordinary)
+            {
+                // Ignore cases where there will already be a warning due to the AttributeTargets.Method usage
+                // restriction on the attribute definition in the framework.
+                if (MethodKind != MethodKind.Constructor && MethodKind != MethodKind.StaticConstructor)
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, arguments.AttributeSyntaxOpt.Location);
+                }
+                return;
+            }
+
+            Debug.Assert(ContainingType is object);
+
+            if (isInaccessible(this) || containingTypes(this).Any(isInaccessible))
+            {
+                var topLevelType = containingTypes(this).Last();
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeAccessibleOutsideTopLevelType, arguments.AttributeSyntaxOpt.Location, Name, topLevelType.Name);
+            }
+
+            if (!IsStatic)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeStatic, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ParameterCount > 0)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotHaveParameters, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (!ReturnsVoid)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustReturnVoid, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (IsGenericMethod)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeGeneric, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            if (ContainingType.IsGenericType)
+            {
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustNotBeContainedInGenericType, arguments.AttributeSyntaxOpt.Location, Name);
+            }
+
+            DeclaringCompilation.AddModuleInitializerMethod(this);
+
+            static bool isInaccessible(Symbol symbol)
```

Comment:
>isInaccessible [](start = 24, length = 14)

We are duplicating the code. I am pretty sure we already have a helper method that we can use to figure out if the method is accessible on the module level.

Reply:
I think this method should work for our scenario:
```
    internal static class AccessCheck
    {
        /// <summary>
        /// Checks if 'symbol' is accessible from within assembly 'within'.
        /// </summary>
        public static bool IsSymbolAccessible(
            Symbol symbol,
            AssemblySymbol within,
            ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```

---
In reply to: [418140520](https://github.com/dotnet/roslyn/pull/43301#discussion_r418140520) [](ancestors = 418140520)

Reply:
Perfect answer to my question, thanks!

Reply:
This instantiates a new hashset some of the time, presumably only when returning `false`. Okay to ignore that it does this?

Reply:
> This instantiates a new hashset some of the time, presumably only when returning false. Okay to ignore that it does this?


The hashset should never be ignored, regardless of the return value. We have a helper to add useSiteDiagnostics to a DiagnosticBag. Look for places in the product where locals of ```HashSet<DiagnosticInfo>``` type are introduced and how they are dealt with afterwards. You might run into a conditional logic somewhere, but we are moving away from that, useSiteDiagnostics should be added to DiagnosticBag unconditionally as long as it is not null.  

---
In reply to: [418366117](https://github.com/dotnet/roslyn/pull/43301#discussion_r418366117) [](ancestors = 418366117)

Reply:
Maybe I should have listed these first, but the overwhelming majority of Find All References on this method showed me by example that the thing to do would be to ignore the hashset. I've actually only found one instance out of 20 usages that I've looked at so far that *doesn't* ignore the hashset. Here are some:

https://github.com/dotnet/roslyn/blob/3328ade9ac311024c0c00de0f29f93a85dc48866/src/Compilers/CSharp/Portable/Binder/Semantics/SemanticFacts.cs#L61-L63

https://github.com/dotnet/roslyn/blob/3328ade9ac311024c0c00de0f29f93a85dc48866/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L2003-L2008

https://github.com/dotnet/roslyn/blob/3328ade9ac311024c0c00de0f29f93a85dc48866/src/Compilers/CSharp/Portable/Symbols/PropertySymbol.cs#L283-L290

https://github.com/dotnet/roslyn/blob/3328ade9ac311024c0c00de0f29f93a85dc48866/src/Compilers/CSharp/Portable/Symbols/OverriddenOrHiddenMembersHelpers.cs#L968

https://github.com/dotnet/roslyn/blame/3328ade9ac311024c0c00de0f29f93a85dc48866/src/Compilers/CSharp/Portable/Binder/Semantics/SemanticFacts.cs#L36-L41

I'm worried that adding diagnostics to the bag would result in duplicate diagnostics. How could I be sure that's not the case?

Reply:
> I'm worried that adding diagnostics to the bag would result in redundant diagnostics. How could I be sure that's not the case?

Do not worry about that. We don't.

> Maybe I should have listed these first, but the overwhelming majority of Find All References on this method showed me by example that the thing to do would be to ignore the hashset. I've actually only found one instance out of 20 usages that I've looked at so far that doesn't ignore the hashset.

All the examples you gave are for scenarios that are not interested in any diagnostics, there is simply no place to report anything for them. I specifically didn't ask you to look for references to IsSymbolAccessible API, but look for something else. 

---
In reply to: [418784053](https://github.com/dotnet/roslyn/pull/43301#discussion_r418784053) [](ancestors = 418784053)

Reply:
> I specifically didn't ask you to look for references to IsSymbolAccessible API, but to look for something else.

I am fully aware, thanks! It would be nice to have some concept of what these diagnostics could look like so that I know where it fits in the picture.

Reply:
@AlekseyTs Updated. Could you help me figure out what a potential diagnostic from this call would look like so I can see how it looks next to module initializer diagnostics?

Reply:
> Could you help me figure out what a potential diagnostic from this call would look like so I can see how it looks next to module initializer diagnostics?

The most common example for use-side diagnostics today is an error about missing metadata reference that might affect or actually affects analysis performed by the compiler. It could also be a reference unification warning, or an error about unsupported metadata that compiler ran into. I do not think that any of those are actually applicable (can be reported) for our scenario, given the inputs that we give to the API. So, it is very unlikely that we get any use-site diagnostic here. However, in the long run it is more robust to not assume that nothing is going to change in the future. The rule we should follow, always report use-site diagnostics, unless there is a real known reason not to. For example, if reporting them breaks a success scenario.   

---
In reply to: [420521639](https://github.com/dotnet/roslyn/pull/43301#discussion_r420521639) [](ancestors = 420521639)

Reply:
Thanks, that's reassuring!